src/Java.Interop.Tools.JavaTypeSystem/Adapters/JavaXmlApiExporter.cs

```diff
@@ -0,0 +1,359 @@
+			//// For such case, it does not skip generation.
+			if (method.IsSynthetic && (method.BaseMethod == null || check (method)))
+				return;
+
+			//// Here we skip most of the overriding methods of a virtual method, unless
+			//// - the method visibility or final-ity has changed: protected Object#clone() is often
+			////   overriden as public. In that case, we need a "new" method.
+			//// - the method is covariant. In that case we need another overload.
+			//// - they differ in "abstract" or "final" method attribute.
+			//// - the derived method is static.
+			//// - the base method is in the NON-public class.
+			//// - none of the arguments are type parameters.
+			//// - finally, it is the synthetic method already checked above.
+			if (method.BaseMethod != null &&
+			    !method.BaseMethod.IsAbstract &&
+			    method.BaseMethod.Visibility == method.Visibility &&
+			    method.BaseMethod.IsAbstract == method.IsAbstract &&
+			    method.BaseMethod.IsFinal == method.IsFinal &&
+			    !method.IsSynthetic &&
+			    check (method))
+				return;
+
+			SaveMember (method, writer, "method",
+				XmlConvert.ToString (method.IsAbstract),
+				XmlConvert.ToString (method.IsNative),
+				GetVisibleReturnTypeString (method),
+				XmlConvert.ToString (method.IsSynchronized),
+				null,
+				null,
+				null,
+				null,
+				null,
+				method.Parameters,
+				method.IsBridge,
+				method.ReturnJni,
+				method.IsSynthetic,
+				method.ReturnNotNull);
+		}
+
+		static string GetVisibleReturnTypeString (JavaMethodModel method)
+		{
+			if (GetVisibleNonSpecialType (method, method.ReturnTypeModel) is JavaTypeReference jtr)
+				return jtr.ToString ();
+
+			return method.Return;
+		}
+
+		public static string? GetVisibleParamterTypeName (this JavaParameterModel parameter)
+		{
+			if (GetVisibleNonSpecialType (parameter.ParentMethod, parameter.TypeModel) is JavaTypeReference jtr)
+				return jtr.ToString ();
+
+			return parameter.GenericType;
+		}
+
+		static JavaTypeReference? GetVisibleNonSpecialType (JavaMethodModel method, JavaTypeReference? r)
+		{
+			if (r == null || r.SpecialName != null || r.ReferencedTypeParameter != null || r.ArrayPart != null)
+				return null;
+
+			var requiredVisibility = method?.Visibility == "public" && method.ParentType?.Visibility == "public" ? "public" : method?.Visibility;
+
+			for (var t = r; t != null; t = (t.ReferencedType as JavaClassModel)?.BaseTypeReference) {
+				if (t.ReferencedType == null)
+					break;
+				if (IsAcceptableVisibility (required: requiredVisibility, actual: t.ReferencedType.Visibility))
+					return t;
+			}
+
+			return null;
+		}
+
+		static bool IsAcceptableVisibility (string? required, string? actual)
+		{
+			if (required == "public")
+				return actual == "public";
+			else
+				return true;
+		}
+
+		static void SaveMember (JavaMemberModel m, XmlWriter writer, string elementName,
+					string? abs, string? native, string? ret, string? sync,
```

Comment:
Only two tab stops are needed here, not three.

…wow, that's a lot of parameters...